### PR TITLE
be more radical by presenting a non-namespaced view-model in starcoun…

### PR DESCRIPTION
**Pull request by warpech**
…ter-include within a red-dotted outline

Screenshots:

- https://drive.google.com/open?id=1di0Toil0QXRoVvIzj1S1JIXwswA5HZdt
- https://drive.google.com/open?id=1z2LvfWOP8GKI_Mq0p2hTbL7sdxw_6kcJ

This was discussed with @mmnosek and @miyconst.

In the beginning I wanted to be more radical by blurring the view or hiding it completely. This was Konstantin's suggestion to use the outline.

### CODE REVIEW

File: starcounter-include.html [hunk 1]
```diff
@@ -462,11 +462,15 @@
...
         function checkForNonNamespaced(viewModel, element){
             if(viewModel && typeof viewModel.Html !== 'undefined'){
-                console.warn(`Your view-model is probably not namespaced!
-`, viewModel, `
-We no longer support those.
-Please make sure it's a result of \`Self.GET\` on serverside. If it's not a result of blending, then replace \`<starcounter-include view-model="{{viewModel}}">\` with  \`<div><imported-template model="{{viewModel}}" href="{{viewModel.Html}}"></imported-template></div>\`.
-`, element);
+                console.warn(`The view ${viewModel.Html} is enclosed in a red-dotted line, because it is incorrectly provided to <starcounter-include> without an app namespace. Your options are: 
+1. Request the blendable Json subtree with Self.GET. Then you can keep to use <starcounter-include> in HTML.
+2. Or, replace <starcounter-include> in HTML with <div><imported-template model="{{viewModel}}" href="{{viewModel.Html}}"></imported-template></div> to insert the Json subtree in a non-blendable fashion. Then you can keep your C# that assigns a Json instance to a parent object without Self.GET.`, viewModel, element);
```
Comment (tomalec) [inline starcounter-include.html]: I'd add a note that `declarative-shadow-dom` will not work in second case. As this was a problem for few devs.

File: starcounter-include.html [hunk 2]
```diff
@@ -462,11 +462,15 @@
...
+                element.setAttribute("style", "outline: 4px red dotted; outline-offset: -4px");
+                element.wasWarned = true;
+            }
+            else if(element.wasWarned) {
```
Comment (tomalec) [inline starcounter-include.html]: What is this supposed to deliver? Flash the warning every second call to `checkForNoNNamespaced` I found it confusing.

File: starcounter-include.html [hunk 3]
```diff
@@ -462,11 +462,15 @@
...
-                console.warn(`Your view-model is probably not namespaced!
-`, viewModel, `
-We no longer support those.
-Please make sure it's a result of \`Self.GET\` on serverside. If it's not a result of blending, then replace \`<starcounter-include view-model="{{viewModel}}">\` with  \`<div><imported-template model="{{viewModel}}" href="{{viewModel.Html}}"></imported-template></div>\`.
-`, element);
+                console.warn(`The view ${viewModel.Html} is enclosed in a red-dotted line, because it is incorrectly provided to <starcounter-include> without an app namespace. Your options are: 
+1. Request the blendable Json subtree with Self.GET. Then you can keep to use <starcounter-include> in HTML.
+2. Or, replace <starcounter-include> in HTML with <div><imported-template model="{{viewModel}}" href="{{viewModel.Html}}"></imported-template></div> to insert the Json subtree in a non-blendable fashion. Then you can keep your C# that assigns a Json instance to a parent object without Self.GET.`, viewModel, element);
+                element.setAttribute("style", "outline: 4px red dotted; outline-offset: -4px");
```
Comment (tomalec) [inline starcounter-include.html]: I'd use `element.style` not to overwrite upcoming FOUC solution.

**Comment by warpech** [comment]
I think/hope that the tests will pass when https://github.com/Starcounter/starcounter-include/pull/111 is merged.

**Review by tomalec — COMMENTED** [review]
If you ask me, I'd go for blurring content. But if you really want to make the result look bad and shout at the dev to change it, I'm fine with this border :)

File: starcounter-include.html (hunk 2 above)
Comment (warpech) [reply starcounter-include.html]: It removes the border when you replace an non-namespaced view-model with a namespaced view-model. 

Why will it flash every second call?

**Comment by warpech** [comment]
Thanks for the review. Blur is too radical, because it makes the page unusable. I'd rather go for annoying than unusable.

I pushed reaction to all comments except https://github.com/Starcounter/starcounter-include/pull/114#discussion_r218102351, to which I wrote an question.

File: starcounter-include.html [hunk 4]
```diff
@@ -462,11 +462,16 @@
...
-`, element);
+                console.warn(`The view ${viewModel.Html} is enclosed in a red-dotted line, because it is incorrectly provided to <starcounter-include> without an app namespace. Your options are: 
+1. Request the blendable Json subtree with Self.GET. Then you can keep to use <starcounter-include> in HTML.
+2. Or, replace <starcounter-include> in HTML with <div><imported-template model="{{viewModel}}" href="{{viewModel.Html}}"></imported-template></div> to insert the Json subtree in a non-blendable fashion. Then you can keep your C# that assigns a Json instance to a parent object without Self.GET. In that case the view does not support declarative-shadow-dom.`, viewModel, element);
+                element.style.outline = "4px red dotted";
+                element.style.outlineOffset = "-4px";
+                element.wasWarned = true;
+            }
+            else if(element.wasWarned) {
+                element.removeAttribute("style");
```
Comment (tomalec) [inline starcounter-include.html]: I'd use `element.style` here as well.

File: starcounter-include.html (hunk 2 above)
Comment (tomalec) [reply starcounter-include.html]: 🤦‍♂️ You are right, my bad.

File: starcounter-include.html [hunk 5]
```diff
@@ -462,11 +462,16 @@
...
-Please make sure it's a result of \`Self.GET\` on serverside. If it's not a result of blending, then replace \`<starcounter-include view-model="{{viewModel}}">\` with  \`<div><imported-template model="{{viewModel}}" href="{{viewModel.Html}}"></imported-template></div>\`.
-`, element);
+                console.warn(`The view ${viewModel.Html} is enclosed in a red-dotted line, because it is incorrectly provided to <starcounter-include> without an app namespace. Your options are: 
+1. Request the blendable Json subtree with Self.GET. Then you can keep to use <starcounter-include> in HTML.
+2. Or, replace <starcounter-include> in HTML with <div><imported-template model="{{viewModel}}" href="{{viewModel.Html}}"></imported-template></div> to insert the Json subtree in a non-blendable fashion. Then you can keep your C# that assigns a Json instance to a parent object without Self.GET. In that case the view does not support declarative-shadow-dom.`, viewModel, element);
+                element.style.outline = "4px red dotted";
+                element.style.outlineOffset = "-4px";
+                element.wasWarned = true;
+            }
+            else if(element.wasWarned) {
```
Comment (miyconst) [inline starcounter-include.html:514]: As far as I remember, we agreed to use default Visual Studio formatting for JavaScript and C#. Thus, please make sure that `Ctrl + K + D` does not change the file.

In this particular case it should be:

```js
} else if (element.wasWarned) {
```

Comment (miyconst) [reply starcounter-include.html:514]: Should we add a [link](https://docs.starcounter.io/topic-guides/blendable-web-apps/view-model-attaching#attachment-rule-components) to `starcounter.io` for more info regarding `Self.GET` and view attaching?

Comment (warpech) [reply starcounter-include.html:514]: Great idea, actually I'll add a link to the blog post that explains the problem.

Regarding the formatting, sorry, I just stick to the formatting that is already in the file.

**Comment by warpech** [comment]
Ready for re-review.

**Review by tomalec — APPROVED** [review]
LGTM, I hope FOUC PR will fix flaky tests

**Pull request merged**